torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -718,6 +733,132 @@ def call_function(
                 return TupleVariable([TupleVariable(items), aux])
 
 
+class FunctorchVmapHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    def call_function(
+        self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
+    ) -> "VariableTracker":
+        from . import ConstantVariable, TensorVariable
+        from .builder import wrap_fx_proxy
+
+        if not torch._dynamo.config.capture_func_transforms:
+            unimplemented("torch.func.vmap capture is disabled")
+
+        checkpoint = tx.copy_graphstate()
+        graph_checkpoint = tx.output.graph
+
+        # unpack args
+        fn = args[0]
+        in_dims = args[1]
+        out_dims = args[2]
+        randomness = args[3]
+        chunk_size = args[4]
+        batch_input_args = args[5:]
+
+        if kwargs:
+            unimplemented(
+                "NYI - torch.func.vmap: kwargs arguments are currently unsupported."
+            )
+
+        if chunk_size.value is not None:
+            unimplemented(
+                "NYI - torch.func.vmap is not implemented when chunk_size is passed"
+            )
+
+        flat_args, arg_spec = torch.utils._pytree.tree_flatten(batch_input_args)
+        in_dims_v = in_dims.as_python_constant()
+        in_dims_v = in_dims_v if isinstance(in_dims_v, int) else list(in_dims_v)
+        broadcasted_in_dims = torch._functorch.vmap._broadcast_to_and_flatten(
+            in_dims_v, arg_spec
+        )
```

Comment:
Hmmm, can we put this into `vmap` and not `vmap_impl`? So that we don't need to duplicate the logic in Dynamo?

Reply:
I don't think dynamo handles `_broadcast_to_and_flatten` yet

```python
def main(foo, in_dims):
    _, in_dims_spec = torch.utils._pytree.tree_flatten(in_dims)
    broadcasted_in_dims = torch.utils._pytree._broadcast_to_and_flatten(in_dims, in_dims_spec)

torch.compile(main, backend='eager', fullgraph=True)(torch.randn(()), 0)

```

Fails with 
```
2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] WON'T CONVERT main /home/kshiteej/Pytorch/pytorch_functorch/test/test_trace_vmap.py line 240 
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] due to: 
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] Traceback (most recent call last):
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING]   File "/home/kshiteej/Pytorch/pytorch_functorch/torch/_dynamo/utils.py", line 1301, in visit
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING]     return n.meta["example_value"]
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] KeyError: example_value
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] 
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] from user code:
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING]    File "/home/kshiteej/Pytorch/pytorch_functorch/test/test_trace_vmap.py", line 242, in main
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING]     broadcasted_in_dims = torch.utils._pytree._broadcast_to_and_flatten(in_dims, in_dims_spec)
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] 
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] 
[2023-08-01 12:31:32,807] torch._dynamo.convert_frame: [WARNING] 
```

Reply:
https://github.com/pytorch/pytorch/issues/106550